src/sensors/elec.rs

```diff
@@ -14,21 +26,142 @@ pub fn filter_pde(pde: f32) -> bool {
+}
+
+/// combine the individual hits into information for each sensor.
+/// Currently uses the minimum as the time for each sensor which
+/// might prove too inflexible. Should be generalised.
+pub fn combine_sensor_hits(event_id: u32, hits: Vec<MCSensorHit>) -> Vec<MCQT> {
+    hits.into_iter()
+        .group_by(|h| h.sensor_id)
+        .into_iter()
+        .map(|(s, grp)| qt_min_time(event_id, s, grp.collect()))
+        .collect()
+}
+
+// This seems like something that would be in a crate somewhere.
+fn dot((x1,y1,z1): (Length, Length, Length), (x2,y2,z2): (Length, Length, Length)) -> Area {
+    x1*x2 + y1*y2 + z1*z2
+}
+
+// Dot product separation for now, need to implement other algorithm(s)
+fn dot_product_clusters(hits: &[MCQT], xyzs: &SensorMap, threshold: f32, min_sensors: usize) -> Option<(Vec<MCQT>, Vec<MCQT>)> {
+    let max_sens = hits.iter().max_by_key(|e| e.q).map(|e| e.sensor_id)?;
+    let &(xm, ym, zm) = xyzs.get(&max_sens)?;
+    let (c1, c2): (Vec<MCQT>, Vec<MCQT>) = hits.iter()
+            .filter(|&MCQT { q, .. }| *q as f32 > threshold)
+            .partition(|&MCQT { sensor_id, .. }| dot((xm, ym, zm), *xyzs.get(sensor_id).unwrap()) > Area::ZERO);
+    if c1.len() >= min_sensors && c2.len() >= min_sensors { Some((c1, c2)) } else { None }
+}
+
+/// Reconstruct LORs based on the barycentre of the 2 selected clusters (dot prod) without DOI.
+fn lor_from_sensor_positions(hits: &[MCQT], xyzs: &SensorMap, threshold: f32, min_sensors: usize) -> Option<Hdf5Lor> {
+    let (c1, c2) = dot_product_clusters(hits, xyzs, threshold, min_sensors)?;
+    // TODO time just taken as minimum of minima. Averaging used elsewhere.
+    let b1 = sipm_charge_barycentre(&c1, xyzs)?;
+    let b2 = sipm_charge_barycentre(&c2, xyzs)?;
+    let q1 = c1.iter().map(|&MCQT { q, .. }| q as f32).sum();
+    let q2 = c2.iter().map(|&MCQT { q, .. }| q as f32).sum();
+    // Just dummy time for now until we work out the best way to sort.
+    Some(Hdf5Lor {
+        dt: ns_(b2.3 - b1.3),
+        x1: mm_(b1.0), y1: mm_(b1.1), z1: mm_(b1.2),
+        x2: mm_(b2.0), y2: mm_(b2.1), z2: mm_(b2.2),
+        q1, q2, E1: f32::NAN, E2: f32::NAN,
+    })
+}
+
+fn sipm_charge_barycentre(hits: &[MCQT], xyzs: &SensorMap) -> Option<(Length, Length, Length, Time)> {
+    let mut qs = Ratio::ZERO;
+    let mut xx = Length::ZERO;
+    let mut yy = Length::ZERO;
+    let mut zz = Length::ZERO;
+    let mut tt = Time::ZERO;
+    for &MCQT{ sensor_id, q, t, .. } in hits {
+        let &(x, y, z) = xyzs.get(&sensor_id)?;
+        let q = ratio(q as f32);
+        qs += q;
+        xx += x * q;
+        yy += y * q;
+        zz += z * q;
+        if t < tt { tt = t};
+    }
+    Some((xx / qs, yy / qs, zz / qs, tt))
+}
+
+
+pub fn lor_reconstruction(first_file: &PathBuf, pde: f32, time_mu: f32, time_sigma: f32, threshold: f32, min_sensors: usize) -> Box<dyn Fn(&PathBuf) -> hdf5::Result<(Vec<Hdf5Lor>, usize)>> {
```

Comment:
+ Why no `uom`?
+ I don't understand the `usize` in `-> Result<_, usize>`. `Err` appears never to be returned from this function.
+ At 190 characters, this signature is begging to be split onto separate lines.

Reply:
> I don't understand the usize in -> Result<_, usize>. Err appears never to be returned from this function.

This is the return type required in `makelors.rs` for this type of function. Since, at least for now, I needed to use it there I just used the same return type.

> At 190 characters, this signature is begging to be split onto separate lines.

I think that happens in a later commit.

Reply:
Ah, OK, I misread the return type. It's 
```rust
hdf5::Result<(XXX, usize)>
```
and I understood
```rust
std::result::Result<XXX, usize>
```
So the `usize` belongs to he happy path, and its meaning is 'the number of LORs being returned'.

Though I'm not sure why we bother with it: isn't it just the `.len()` of the vector being returned?

Reply:
Right, the `usize` is the number of events that were processed, which is different from the number of events that produced a LOR. I'll change the variable names to make this clearer, and push to `master` shortly.

Reply:
It should make it to `master` sometime today. The consequences for your code will be that you'll have to change

```diff
-                        ) -> Box<dyn Fn(&PathBuf) -> hdf5::Result<(Vec<Hdf5Lor>, usize)>> {
+                        ) -> Box<dyn Fn(&PathBuf) -> hdf5::Result<LorBatch>> {
```
```diff
-        Ok((lors_from(&events, |evs| lor_from_sensor_positions(evs, &xyzs, threshold, min_sensors, min_charge)), events.len()))
+        Ok(LorBatch::new(lors_from(&events, |evs| lor_from_sensor_positions(evs, &xyzs, threshold, min_sensors, min_charge)), events.len()))
     })
```